pvlib/iotools/bsrn.py

```diff
@@ -4,139 +4,449 @@
+                dfi, metadata = parse_bsrn(io.StringIO(text), logical_records)
+                dfs.append(dfi)
+            # FTP client raises an error if the file does not exist on server
+            except ftplib.error_perm as e:
+                if str(e) == '550 Failed to open file.':
+                    non_existing_files.append(filename)
+                else:
+                    raise ftplib.error_perm(e)
+        ftp.quit()  # Close and exit FTP connection
 
+    # Raise user warnings
+    if not dfs:  # If no files were found
+        warnings.warn('No files were available for the specified timeframe.')
+    elif non_existing_files:  # If only some files were missing
+        warnings.warn(f'The following files were not found: {non_existing_files}')  # noqa: E501
+
+    # Concatenate monthly dataframes to one dataframe
+    if len(dfs):
+        data = pd.concat(dfs, axis='rows')
+    else:  # Return empty dataframe
+        data = _empty_dataframe_from_logical_records(logical_records)
+        metadata = {}
+    # Return dataframe and metadata (metadata belongs to last available file)
+    return data, metadata
+
+
+def parse_bsrn(fbuf, logical_records=['0100']):
+    """
+    Parse a file-like buffer of a BSRN station-to-archive file.
 
     Parameters
     ----------
-    filename: str
-        A relative or absolute file path.
+    fbuf: file-like buffer
+        Buffer of a BSRN station-to-archive data file
+    logical_records: str or list, default: ['0100']
+        List of the logical records (LR) to parse. Options are: '0100', '0300',
+        and '0500'.
+
+    Returns
+    -------
+    data: DataFrame
+        timeseries data from the BSRN archive, see
+        :func:`pvlib.iotools.read_bsrn` for fields.
+    metadata: dict
+        Dictionary containing metadata (primarily from LR0004).
+
+    See Also
+    --------
+    pvlib.iotools.read_bsrn, pvlib.iotools.get_bsrn
+
+    """
+    # Parse metadata
+    fbuf.readline()  # first line should be *U0001, so read it and discard
+    date_line = fbuf.readline()  # second line contains important metadata
+    start_date = pd.Timestamp(year=int(date_line[7:11]),
+                              month=int(date_line[3:6]), day=1,
+                              tz='UTC')  # BSRN timestamps are UTC
+
+    metadata = {}  # Initilize dictionary containing metadata
+    metadata['start date'] = start_date
+    metadata['station identification number'] = int(date_line[:3])
+    metadata['version of data'] = int(date_line.split()[-1])
+    for line in fbuf:
+        if line[2:6] == '0004':  # stop once LR0004 has been reached
+            break
+        elif line == '':
+            raise ValueError('Mandatory record LR0004 not found.')
+    metadata['date when station description changed'] = fbuf.readline().strip()
+    metadata['surface type'] = int(fbuf.readline(3))
+    metadata['topography type'] = int(fbuf.readline())
+    metadata['address'] = fbuf.readline().strip().strip()

```

Comment:
More for my own education: what does the second `strip()` accomplish here?

Reply:
Well, the second strip() is redundant. But strip() removes starting and trailing whitespaces, as well as "\r\n" and "\n"